### PR TITLE
Add Raven features

### PR DESCRIPTION
**Pull request by rhine3**
Main changes to be aware of:

* Large changes to `opensoundscape/raven.py` and `tests/test_raven.py`
  * Remove default column names in many functions (defaults were unnecessary and led to confusing silent errors)
  * Add helper function, `raven.split_starts_ends()` which splits a Raven annotation file based on lists of starting and ending times for each clip.
  * Make `raven.split_single_annotation()` rely on `raven.split_starts_ends()` and give it the ability to calculate overlap
  * Create new function, `raven.raven_audio_split_and_save()` which finds and simultaneously splits raven files and matching audio files.
* Moderate changes to `opensoundscape/audio.py`
  * Make `audio.Audio.split()` logic more elegant (changes made after thinking through logic for `raven.split_single_annotation()`)
  * In `audio.split_and_save()`, put clip names into an index instead of a column
* Rearranged the organization of supplementary files (e.g. wavs) in the `tests/` folder 
* Made small changes to many of the `.ipynb`s and tests to account for rearranged supplementary files

### CODE REVIEW

File: opensoundscape/audio.py [hunk 1]
```diff
@@ -266,79 +266,69 @@ def split(self, clip_duration, clip_overlap=0, final_clip=None):
...
-                                - "full":               Increase the overlap to yield a clip with clip_duration
-                                - "extend":             Similar to remainder but extend the clip to clip_duration
+            clip_duration (float):  The duration in seconds of the clips
+            clip_overlap (float):   The overlap of the clips in seconds [default: 0]
+            final_clip (str):       Behavior if final_clip is less than clip_duration seconds long. [default: None]
+                By default, ignores final clip entirely.
+                Possible options (any other input will ignore the final clip entirely),
+                    - "remainder":  Include the remainder of the Audio (clip will not have clip_duration length)
+                    - "full":       Increase the overlap to yield a clip with clip_duration length
+                    - "extend":     Similar to remainder but extend (repeat) the clip to reach clip_duration length
```
Comment (sammlapp) [inline opensoundscape/audio.py:276]: I thought the extend behavior would extend the clip with silence - didn't it used to behave like that? 
Extend function is misleading - Maybe we should make Audio.extend() add silence and call the current function audio.repeat() or something.

File: opensoundscape/audio.py [hunk 2]
```diff
@@ -349,39 +339,48 @@ def split_and_save(
...
         )
-        clip["filename"] = clip_name
+        clip_names.append(clip_name)
         if not dry_run:
             clip["clip"].save(clip_name)
 
     # Convert [{k: v}] -> {k: [v]}
-    return pd.DataFrame(
-        {key: [clip[key] for clip in clips] for key in clips[0].keys() if key != "clip"}
+    sels = pd.DataFrame(
```
Comment (sammlapp) [inline opensoundscape/audio.py:378]: what is "sels"?

File: opensoundscape/audio.py (hunk 1 above)
Comment (rhine3) [reply opensoundscape/audio.py:276]: To the best of my knowledge, we always used "extend" to mean "loop the recording so it extends to the desired length" -- I think this was modeled after a Lasseck approach

+1 to the idea of changing `Audio.extend()` to either `Audio.repeat()`  or `Audio.loop()` then adding an "extend with silence" method

File: opensoundscape/audio.py (hunk 2 above)
Comment (rhine3) [reply opensoundscape/audio.py:379]: I think "sels" was supposed to mean "selections." The return is identical to the return that was already there, just formatted by naming the variable first. I'm not sure why I did this, though, and the name doesn't make sense, so I'll change it.

File: opensoundscape/audio.py [hunk 3]
```diff
@@ -137,8 +137,28 @@ def from_bytesio(
...
+            a new Audio object of the desired sample rate
+        """
+        samples_resampled = librosa.resample(
+            self.samples, orig_sr=self.sample_rate, target_sr=sample_rate
+        )
+
+        return Audio(
+            samples_resampled,
+            sample_rate,
+            resample_type=self.resample_type,
```
Comment (sammlapp) [inline opensoundscape/audio.py:156]: resample type refers to the algorithm for resampling. So maybe the librosa.resample call should use the audio object's self.resample_type - otherwise, including self.resample_type in the new object is somewhat misleading. Resample-type could (should?) also be an optional argument to .resample()

File: opensoundscape/audio.py [hunk 4]
```diff
@@ -266,79 +286,71 @@ def split(self, clip_duration, clip_overlap=0, final_clip=None):
...
+        # Now we have the starts and ends
+        final_idx = len(ends) - 1
+        to_return = [None] * (final_idx + 1)
+        for idx, (start, end) in enumerate(zip(starts, ends)):
+            # By default
+            begin_time = start
+            end_time = end
+
+            # Change defaults to handle final clip
+            if idx >= final_idx:
```
Comment (sammlapp) [inline opensoundscape/audio.py:324]: since this is a very critical feature - are you sure that it behaves the same way as the old code? Or if there are differences in how it handles corner cases, what are they?

File: opensoundscape/raven.py [hunk 5]
```diff
@@ -159,51 +163,56 @@ def query_annotations(directory, cls, col="class", print_out=False):
     return output
 
 
-def split_single_annotation(
-    raven_path, split_len_s, total_len_s=None, species=None, col="class"
-):
-    """Split a Raven selection table into short annotations
+def split_starts_ends(raven_file, col, starts, ends, species=None, min_label_len=0):
+    """Use a list of start and end times to split a Raven files
```
Comment (sammlapp) [inline opensoundscape/raven.py:167]: does this function aggregate annotations (from one column of a Raven annotation file) over each start-end time segment into one-hot labels? It's hard to tell from the docstring

File: opensoundscape/raven.py [hunk 6]
```diff
@@ -159,51 +163,56 @@ def query_annotations(directory, cls, col="class", print_out=False):
...
+    if species is None:
         species = selections_df[col].unique()
+    elif type(species) == str:
+        species = [species]
+    species.sort()
 
     cols = ["seg_start", "seg_end", *species]
     splits_df = pd.DataFrame(columns=cols)
 
     # Create a dataframe of split_len_s segments and the annotations in each segment
```
Comment (sammlapp) [inline opensoundscape/raven.py:204]: outdated comment?

File: opensoundscape/raven.py [hunk 7]
```diff
@@ -212,13 +221,105 @@ def split_single_annotation(
...
+    raven_file,
+    col,
+    split_len_s,
+    overlap_len_s=0,
+    total_len_s=None,
+    keep_final=False,
+    species=None,
+    min_label_len=0,
+):
+    """Split a Raven selection table into short annotations
```
Comment (sammlapp) [inline opensoundscape/raven.py:239]: does this result in one-hot annotations for even-lengthed time segments? 

maybe could say "Aggregate one-hot annotations for even-lengthed time segments, drawing annotations from a specified column of a Raven selection table"

File: tests/raven/raven_annots/MSD-0003_20180427_5minstart00-15.Table.1.selections.txt [hunk 8]
```diff
@@ -0,0 +1,142 @@
+Selection	View	Channel	Begin Time (s)	End Time (s)	Low Freq (Hz)	High Freq (Hz)	Species
```
Comment (sammlapp) [inline tests/raven/raven_annots/MSD-0003_20180427_5minstart00-15.Table.1.selections.txt:1]: should this file still exist? I assume it is created by a test

**Review by sammlapp — APPROVED** [review]

File: opensoundscape/audio.py (hunk 1 above)
Comment (rhine3) [reply opensoundscape/audio.py:296]: Added an enhancement request for this (#265) but will not change in this PR

File: opensoundscape/audio.py (hunk 2 above)
Comment (rhine3) [reply opensoundscape/audio.py]: Fixed this

File: opensoundscape/audio.py (hunk 3 above)
Comment (rhine3) [reply opensoundscape/audio.py]: Added a `resample_type` argument that by default is `self.resample_type` of the original `Audio` instance. The `resample_type` used in this method is used to initialize the new Audio object.

File: opensoundscape/audio.py (hunk 4 above)
Comment (rhine3) [reply opensoundscape/audio.py:332]: I am pretty confident this code behaves the same as the old code. It passes all the tests that the previous split code passed, plus a bunch of new splitting tests I added to `test_audio.py`. And vice versa, when I apply the new tests to the old splitting code, the old code passes all of the new tests except for a test that catches the corner case described below.

The one corner case that differs is what I thought was unexpected/undesirable behavior in the previous code: when using `final_clip="full"`, the old split code casts the start time of the final clip as an integer. This means that when making non-overlapping 5s clips from an Audio object that is, say, 8.2s long, the old code extracted the final clip at (3s, 8s) instead of (3.2s, 8.2s). If for some reason that's the desired behavior, we can change it back pretty easily by changing this line back to `begin_time = int(duration - clip_duration)`: https://github.com/kitzeslab/opensoundscape/blob/0316ba06ba535fba62d5429b7c62feb67f160941/opensoundscape/audio.py#L337

File: tests/raven/raven_annots/MSD-0003_20180427_5minstart00-15.Table.1.selections.txt (hunk 8 above)
Comment (rhine3) [reply tests/raven/raven_annots/MSD-0003_20180427_5minstart00-15.Table.1.selections.txt:1]: Yep, this file was added to test `raven.py`'s splitting behavior on a longer/more complicated splitting file

File: opensoundscape/raven.py (hunk 5 above)
Comment (rhine3) [reply opensoundscape/raven.py]: Good suggestion; added more descriptive docstring

File: opensoundscape/raven.py (hunk 6 above)
Comment (rhine3) [reply opensoundscape/raven.py]: Good catch; fixed this up

File: opensoundscape/raven.py (hunk 7 above)
Comment (rhine3) [reply opensoundscape/raven.py:247]: Good suggestion; added your suggested text to the docstring
